### PR TITLE
[#5255][autodeploy] Update FuseAllreduceResidualRMSNorm to use pattern matcher utility; remove fuse_collective

### LINKED ISSUE #5255: Migrate other transformations to use torch._inductor pattern matcher
Separate `fuse_allreduce_residual_rmsnorm` and `fuse_collectives` into pattern matching and fusion phase, then apply torch._inductor pattern matcher for pattern matching

### PR DESCRIPTION
**Pull request by Fridah-nv**
## Summary by CodeRabbit

- Refactor
  - Switched to pattern-based fusion for all-reduce + residual + RMSNorm, improving reliability and coverage of fusions and consistent reporting of applied fusions.
- Bug Fixes
  - Removed a conflicting collective fusion option from the default configuration to avoid unintended behavior.
  - Corrected selection of the dim-1 distributed reduce operator during sharding to improve compatibility and prevent execution errors.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [X] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/utils/sharding_utils.py [hunk 1]
```diff
@@ -169,7 +169,7 @@ def set_new_param(submod: nn.Module, param_key: str, remove: bool = False) -> to
     # figure out the right dist op
     dist_lookup = {
         0: (torch.ops.auto_deploy.torch_dist_all_gather, -1),
-        1: (torch.ops.auto_deploy.torch_dist_all_reduce,),
+        1: (torch.ops.auto_deploy.torch_dist_all_reduce.default,),
```
Comment (Fridah-nv) [inline tensorrt_llm/_torch/auto_deploy/utils/sharding_utils.py:172]: Need to dispatch the implementation to `torch_dist_all_reduce.default` because torch.export will always dispatch to OpOverLoadPacket like `default`. Our pattern matcher uses torch.export as export function and expects the graph to match uses the same OpOverLoadPacket

Comment (Fridah-nv) [reply tensorrt_llm/_torch/auto_deploy/utils/sharding_utils.py:172]: In the future we can have another pass that dispatch our manually added ops to `.default` as @lucaslie suggested.

File: tensorrt_llm/_torch/auto_deploy/transform/interface.py [hunk 2]
```diff
@@ -328,11 +342,16 @@ def _run_pre_cleanup(self, gm: GraphModule, info: TransformInfo) -> Tuple[bool,
...
 
+        use_meta = not self._has_finished_stage(gm, Stages.WEIGHT_LOAD)
+
         # check if run cleanup depending on the config and info
         if self.config.requires_shape_prop and not has_valid_shapes:
             self._log_info("running pre-cleanup with shape_prop")
             canonicalize_graph(gm)
-            with lift_to_meta(gm):
+            if use_meta:
+                with lift_to_meta(gm):
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/transform/interface.py:352]: how about we just check for the device on the parameters?

If it's meta we lift to meta, otherwise we just run it?

File: tensorrt_llm/_torch/auto_deploy/transform/interface.py [hunk 3]
```diff
@@ -328,11 +342,16 @@ def _run_pre_cleanup(self, gm: GraphModule, info: TransformInfo) -> Tuple[bool,
...
+                    run_shape_prop(gm)
+            else:
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/transform/interface.py:354]:
```suggestion
            with lift_to_meta(gm) if use_meta else nullcontext():
```

File: tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_allreduce_residual_rmsnorm_fusion.py [hunk 4]
```diff
@@ -35,8 +35,8 @@ def __init__(self, hidden_size, dtype):
         self.norm = RMSNorm(hidden_size, 1e-5, dtype)
 
     def forward(self, x, residual):
-        x = torch.ops.auto_deploy.torch_dist_all_reduce(x)
-        y = x + residual
+        x = torch.ops.auto_deploy.torch_dist_all_reduce.default(x)
+        y = residual + x
```
Comment (lucaslie) [inline tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_allreduce_residual_rmsnorm_fusion.py:39]: any reason why `y = x + residual` doesn't work anymore?

File: tensorrt_llm/_torch/auto_deploy/transform/library/fusion.py [hunk 5]
```diff
@@ -67,7 +62,8 @@ def split_output(tensor: torch.Tensor) -> Tuple[torch.Tensor, ...]:
     # add new linear node + split node
     with gm.graph.inserting_before(linear_nodes[0]):
         fused_linear_node = gm.graph.call_function(
-            get_op_overload_packet(linear_nodes[0].target),
+            # get_op_overload_packet(linear_nodes[0].target),
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/transform/library/fusion.py:65]: nit: remove commented out line

File: tensorrt_llm/_torch/auto_deploy/config/default.yaml [hunk 6]
```diff
@@ -92,13 +93,16 @@ transforms:
...
   fuse_fp4_gemms:
     stage: post_load_fusion
-    enabled: false # TODO: https://github.com/NVIDIA/TensorRT-LLM/issues/4674 this is causing OOMs
+    enabled: true # TODO: https://github.com/NVIDIA/TensorRT-LLM/issues/4674 this is causing OOMs
+    run_shape_prop: true
   fuse_fp8_gemms:
     stage: post_load_fusion
-    enabled: false # TODO: https://github.com/NVIDIA/TensorRT-LLM/issues/4674 this is causing OOMs
+    enabled: true # TODO: https://github.com/NVIDIA/TensorRT-LLM/issues/4674 this is causing OOMs
+    run_shape_prop: true
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/config/default.yaml:105]: I am assuming we gonna keep it at `enabled: False`, right?

File: tensorrt_llm/_torch/auto_deploy/config/default.yaml [hunk 7]
```diff
@@ -92,13 +93,16 @@ transforms:
   ############################################################################################
   fuse_gemms:
     stage: post_load_fusion
-    enabled: false # TODO: https://github.com/NVIDIA/TensorRT-LLM/issues/4674 this is causing OOMs
+    enabled: true # TODO: https://github.com/NVIDIA/TensorRT-LLM/issues/4674 this is causing OOMs
+    run_shape_prop: true
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/config/default.yaml:97]: `run_shape_prop` is expensive. Let's be a little more rigorous when we apply it.

My suggestion: set `requires_shape_prop: true` for `fuse_allreduce_residual_rmsnorm`. Then it's run as pre-cleanup steam for `fuse_allreduce_residual_rmsnorm` instead of as post-clean up step for every fusion before

File: tensorrt_llm/_torch/auto_deploy/transform/interface.py (hunk 2 above)
Comment (Fridah-nv) [reply tensorrt_llm/_torch/auto_deploy/transform/interface.py]: Thanks for the suggestion, added the utility to check meta tensors of a model.

File: tensorrt_llm/_torch/auto_deploy/config/default.yaml (hunk 6 above)
Comment (Fridah-nv) [reply tensorrt_llm/_torch/auto_deploy/config/default.yaml]: Yes, this is only for testing.

File: tensorrt_llm/_torch/auto_deploy/transform/interface.py (hunk 2 above)
Comment (Fridah-nv) [reply tensorrt_llm/_torch/auto_deploy/transform/interface.py]: Rethinking about this... I think the condition to use `lift_to_meta` for shape propagation is that all inputs to the graph are expected to be on meta device. This includes all parameters, buffers and the placeholder nodes. So we should also check placeholder nodes' device here.
I see a failure in the `match_repeat_kv` unit test that the model and params are all on cuda, but placeholder node is on meta (by the nature of torch.export), and shape propagation fails because of device discrepancy.

File: tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_allreduce_residual_rmsnorm_fusion.py (hunk 4 above)
Comment (lucaslie) [reply tests/unittest/_torch/auto_deploy/unit/multigpu/transformations/library/test_allreduce_residual_rmsnorm_fusion.py:39]: seems useful to have :) Easy to get tripped up otherwise and trivial to add another pattern right?

File: tensorrt_llm/_torch/auto_deploy/transform/interface.py (hunk 2 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/transform/interface.py]: I see you have two versions now, which is the right version?

File: tensorrt_llm/_torch/auto_deploy/transform/interface.py [hunk 8]
```diff
@@ -328,11 +335,13 @@ def _run_pre_cleanup(self, gm: GraphModule, info: TransformInfo) -> Tuple[bool,
         is_clean = info.is_clean
         has_valid_shapes = is_clean and info.has_valid_shapes
 
+        use_meta = isinstance(gm, GraphModule) and has_any_meta_tensor(gm)
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/transform/interface.py:338]: `has_any_meta_tensor` or `placeholders_on_meta`?

Comment (Fridah-nv) [reply tensorrt_llm/_torch/auto_deploy/transform/interface.py:338]: I want to use `placeholders_on_meta `

Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/transform/interface.py:338]: Go for it 👍

**Review by lucaslie — APPROVED** [review]
lgtm
